state/oracledatabase/oracledatabaseaccess.go

```diff
@@ -0,0 +1,377 @@
+	if val, ok := o.metadata.Properties[oracleWalletLocationKey]; ok && val != "" {
+		o.connectionString += "?TRACE FILE=trace.log&SSL=enable&SSL Verify=false&WALLET=" + url.QueryEscape(val)
+	}
+	db, err := sql.Open("oracle", o.connectionString)
+	if err != nil {
+		o.logger.Error(err)
+
+		return err
+	}
+
+	o.db = db
+
+	if pingErr := db.Ping(); pingErr != nil {
+		return pingErr
+	}
+	err = o.ensureStateTable(tableName)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Set makes an insert or update to the database.
+func (o *oracleDatabaseAccess) Set(req *state.SetRequest) error {
+	return state.SetWithOptions(o.setValue, req)
+}
+
+func parseTTL(requestMetadata map[string]string) (*int, error) {
+	if val, found := requestMetadata[metadataTTLKey]; found && val != "" {
+		parsedVal, err := strconv.ParseInt(val, 10, 0)
+		if err != nil {
+			return nil, fmt.Errorf("error in parsing ttl metadata : %w", err)
+		}
+		parsedInt := int(parsedVal)
+
+		return &parsedInt, nil
+	}
+
+	return nil, nil
+}
+
+// setValue is an internal implementation of set to enable passing the logic to state.SetWithRetries as a func.
+func (o *oracleDatabaseAccess) setValue(req *state.SetRequest) error {
+	o.logger.Debug("Setting state value in OracleDatabase")
+	err := state.CheckRequestOptions(req.Options)
+	if err != nil {
+		return err
+	}
+
+	if req.Key == "" {
+		return fmt.Errorf("missing key in set operation")
+	}
+
+	if v, ok := req.Value.(string); ok && v == "" {
+		return fmt.Errorf("empty string is not allowed in set operation")
+	}
+	if req.Options.Concurrency == state.FirstWrite && (req.ETag == nil || len(*req.ETag) == 0) {
+		o.logger.Debugf("when FirstWrite is to be enforced, a value must be provided for the ETag")
+		return fmt.Errorf("when FirstWrite is to be enforced, a value must be provided for the ETag")
+	}
+	var ttlSeconds int
+	ttl, ttlerr := parseTTL(req.Metadata)
+	if ttlerr != nil {
+		return fmt.Errorf("error in parsing TTL %w", ttlerr)
+	}
+	if ttl != nil {
+		if *ttl == -1 {
+			o.logger.Debugf("TTL is set to -1; this means: never expire. ")
+		} else {
+			if *ttl < -1 {
+				return fmt.Errorf("incorrect value for %s %d", metadataTTLKey, *ttl)
+			}
+			ttlSeconds = *ttl
+		}
+	}
+	requestValue := req.Value
+	byteArray, isBinary := req.Value.([]uint8)
+	binaryYN := "N"
+	if isBinary {
+		requestValue = base64.StdEncoding.EncodeToString(byteArray)
+		binaryYN = "Y"
+	}
+
+	// Convert to json string.
+	bt, _ := utils.Marshal(requestValue, json.Marshal)
+	value := string(bt)
+
+	var result sql.Result
+	var tx *sql.Tx
+	if o.tx == nil { // not joining a preexisting transaction.
+		tx, err = o.db.Begin()
+		if err != nil {
+			return fmt.Errorf("failed to start database transaction : %w", err)
+		}
+	} else { // join the transaction passed in.
+		tx = o.tx
+	}
+	etag := uuid.New().String()
+	// QUESTION: only check for etag if FirstWrite specified - or always when etag is supplied??
+	if req.Options.Concurrency != state.FirstWrite {
```

Comment:
/cc @artursouza

Reply:
@daixiang0  @artursouza  this has been resolved as per Discord message: 

Only check for etag if FirstWrite specified - as per Discord message thread https://discord.com/channels/778680217417809931/901141713089863710/938520959562952735. 

I will update the code and extend the integration and unit test to this effect. Commit to follow shortly.

Reply:
committed updated code and integration test